datadog-trace-utils/src/span/mod.rs

```diff
@@ -79,340 +80,94 @@ impl SpanText for BytesString {
-where
-    T: SpanText,
-{
-    pub service: T,
-    pub name: T,
-    pub resource: T,
-    #[serde(skip_serializing_if = "is_empty_str")]
-    pub r#type: T,
-    pub trace_id: u64,
-    pub span_id: u64,
-    #[serde(skip_serializing_if = "is_default")]
-    pub parent_id: u64,
-    pub start: i64,
-    pub duration: i64,
-    #[serde(skip_serializing_if = "is_default")]
-    pub error: i32,
-    #[serde(skip_serializing_if = "HashMap::is_empty")]
-    pub meta: HashMap<T, T>,
-    #[serde(skip_serializing_if = "HashMap::is_empty")]
-    pub metrics: HashMap<T, f64>,
-    // TODO: APMSP-1941 - Replace `Bytes` with a wrapper that borrows the underlying
-    // slice and serializes to bytes in MessagePack.
-    #[serde(skip_serializing_if = "HashMap::is_empty")]
-    pub meta_struct: HashMap<T, Bytes>,
-    #[serde(skip_serializing_if = "Vec::is_empty")]
-    pub span_links: Vec<SpanLink<T>>,
-    #[serde(skip_serializing_if = "Vec::is_empty")]
-    pub span_events: Vec<SpanEvent<T>>,
+pub trait SpanBytes: Debug + Eq + Hash + Borrow<[u8]> + Serialize + Default + Clone {
+    fn from_static_bytes(value: &'static [u8]) -> Self;
 }
 
-/// The generic representation of a V04 span link.
-/// `T` is the type used to represent strings in the span link.
-#[derive(Clone, Debug, Default, PartialEq, Serialize)]
-pub struct SpanLink<T>
-where
-    T: SpanText,
-{
-    pub trace_id: u64,
-    pub trace_id_high: u64,
-    pub span_id: u64,
-    #[serde(skip_serializing_if = "HashMap::is_empty")]
-    pub attributes: HashMap<T, T>,
-    #[serde(skip_serializing_if = "is_empty_str")]
-    pub tracestate: T,
-    #[serde(skip_serializing_if = "is_default")]
-    pub flags: u32,
-}
-
-/// The generic representation of a V04 span event.
-/// `T` is the type used to represent strings in the span event.
-#[derive(Clone, Debug, Default, PartialEq, Serialize)]
-pub struct SpanEvent<T>
-where
-    T: SpanText,
-{
-    pub time_unix_nano: u64,
-    pub name: T,
-    #[serde(skip_serializing_if = "HashMap::is_empty")]
-    pub attributes: HashMap<T, AttributeAnyValue<T>>,
+impl SpanBytes for &[u8] {
+    fn from_static_bytes(value: &'static [u8]) -> Self {
+        value
+    }
 }
 
-#[derive(Clone, Debug, PartialEq)]
-pub enum AttributeAnyValue<T>
-where
-    T: SpanText,
-{
-    SingleValue(AttributeArrayValue<T>),
-    Array(Vec<AttributeArrayValue<T>>),
+impl SpanBytes for Bytes {
+    fn from_static_bytes(value: &'static [u8]) -> Self {
+        Bytes::from_static(value)
+    }
 }
 
-#[derive(Serialize)]
-struct ArrayValueWrapper<'a, T: SpanText> {
-    values: &'a Vec<AttributeArrayValue<T>>,
-}
+/// Trait representing a tuple of (Text, Bytes) types used for different underlying data structures.
+/// Note: The functions are internal to the msgpack decoder and should not be used directly: they're
+/// only exposed here due to the inavailability of min_specialization in stable Rust.
+pub trait TraceData: Default + Debug + Clone + PartialEq + Serialize {
```

Comment:
Why the `Clone` and `PartialEq`?

Reply:
PartialEq is necessary because v04::Span is PartialEq. And yeah, generic types need to have all these derives of the context they are used in, otherwise #[derive] doesn't work.
I have no idea though why I added Clone. Removing.

Reply:
Yeah, cfg_attr doesn't work anyway, that'd need at least a feature flag. That's ugly.